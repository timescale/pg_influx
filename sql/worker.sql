CREATE SCHEMA db_worker;
CREATE TABLE db_worker.disk(_time timestamptz, host text, device text, _tags jsonb, _fields jsonb);
CREATE TABLE db_worker.system(_time timestamp, host text, uptime int, _tags jsonb, _fields jsonb);

CREATE EXTENSION influx WITH SCHEMA db_worker;

\set VERBOSITY terse
\x on
SELECT pg_sleep(1), pid AS worker_pid FROM db_worker.worker_launch('db_worker', 4711::text) AS pid \gset
CALL db_worker.send_packet('cpu,cpu=cpu0,host=fury usage_system=2.0408163264927324,usage_user=2.0408163264927324 1574753954000000000', 4711::text);
CALL db_worker.send_packet('cpu,cpu=cpu1,host=fury usage_system=3.921568627286635,usage_user=3.9215686274649673 1574753954000000000', 4711::text);
CALL db_worker.send_packet('disk,device=nvme0n1p2,fstype=ext4,host=fury,mode=rw,path=/ free=912578965504i,total=1006530654208i,used=42751348736i,used_percent=4.475033200428718 1574753954000000000', 4711::text);
CALL db_worker.send_packet('disk,device=nvme0n1p1,fstype=vfat,host=fury,mode=rw,path=/boot/efi free=527806464i,total=535805952i,used=7999488i,used_percent=1.4929822952022749 1574753954000000000', 4711::text);
CALL db_worker.send_packet('system,host=fury load1=2.13,load15=0.84,load5=1.18,n_cpus=8i,n_users=1i 1574753954000000000', 4711::text);
CALL db_worker.send_packet('system,host=fury uptime=607641i 1574753954000000000', 4711::text);
CALL db_worker.send_packet('system,host=fury uptime_format="7 days,  0:47" 1574753954000000000', 4711::text);
SELECT pg_sleep(2);

SELECT * FROM db_worker.cpu;	--Automatically created
SELECT * FROM db_worker.disk;
SELECT * FROM db_worker.system;

SELECT count(*) FROM pg_stat_activity WHERE pid = :worker_pid;
-- Syntax error, but the worker should not stop
CALL db_worker.send_packet('system,host=fury', 4711::text);
SELECT pg_sleep(1);
SELECT count(*) FROM pg_stat_activity WHERE pid = :worker_pid;

SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE backend_type like '%Influx%';

DROP EXTENSION influx;
DROP TABLE db_worker.cpu;
DROP TABLE db_worker.disk;
DROP TABLE db_worker.system;
DROP SCHEMA db_worker;
