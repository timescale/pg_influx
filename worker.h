/*
 * Copyright 2022 Timescale Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef WORKER_H_
#define WORKER_H_

#include <postgres.h>

#include <funcapi.h>
#include <miscadmin.h>
#include <postmaster/bgworker.h>

#define INFLUX_LIBRARY_NAME "influx"
/* MTU for the wireless interface. We don't bother about digging up
 * the actual MTU here and just pick something that is common. */
#define INFLUX_MTU 1500

typedef struct WorkerArgs {
  char role[32];
  char namespace[32];
  char database[32];
  char service[32];
} WorkerArgs;

void InfluxWorkerInit(BackgroundWorker *worker, WorkerArgs *args);

void PGDLLEXPORT InfluxWorkerMain(Datum dbid) pg_attribute_noreturn();

void InfluxWorkerSetup(WorkerArgs *args);
void InfluxProcessPacket(char *buffer, size_t bytes);

/*
 * Variables that are shared between all workers.
 */
extern volatile sig_atomic_t ReloadConfig;
extern volatile sig_atomic_t ShutdownWorker;
extern Oid NamespaceOid;

#endif /* WORKER_H_ */
