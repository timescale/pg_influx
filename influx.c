/*
 * Copyright 2022 Timescale Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "influx.h"

#include <postgres.h>
#include <fmgr.h>

#include <catalog/namespace.h>
#include <catalog/pg_type.h>
#include <executor/spi.h>
#include <funcapi.h>
#include <miscadmin.h>
#include <postmaster/bgworker.h>
#include <utils/acl.h>
#include <utils/builtins.h>
#include <utils/guc.h>
#include <utils/jsonb.h>
#include <utils/timestamp.h>

#include <stdbool.h>
#include <string.h>

#include "guc.h"
#include "ingest.h"
#include "udp.h"
#include "worker.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(parse_influx);

void PGDLLEXPORT _PG_init(void);

/** Parser state setup. */
IngestState *ParseInfluxSetup(char *buffer) {
  IngestState *state = palloc(sizeof(IngestState));
  IngestStateInit(state, buffer);
  return state;
}

/**
 * Parse one line of data from the parser state and store it as a heap tuple
 * with the columns:
 *
 * 1. The metric
 * 2. The timestamp
 * 3. The tags as a JSONB value
 * 4. The fields as a JSONB value
 *
 * @param state Parser state, with line information.
 * @param tupdesc Tuple descriptor for the data.
 * @returns tuple Heap tuple
 */
static HeapTuple ParseInfluxNextTuple(IngestState *state,
                                      AttInMetadata *attinmeta) {
  int metric_attnum;
  Datum *values;
  bool *nulls;
  Oid *argtypes;
  TupleDesc tupdesc = attinmeta->tupdesc;

  nulls = (bool *)palloc0(tupdesc->natts * sizeof(bool));
  values = (Datum *)palloc0(tupdesc->natts * sizeof(Datum));
  argtypes = (Oid *)palloc0(tupdesc->natts * sizeof(Oid));

  /* Read lines until we find one that can be used. If none are found, we're
   * done. */
  do {
    if (!IngestReadNextLine(state))
      return NULL;
  } while (!CollectValues(&state->metric, attinmeta, argtypes, values, nulls));

  /* This assumes that the metric is a text column. We should probably add a
   * check here, or call the input function for the column type. */
  metric_attnum = SPI_fnumber(tupdesc, "_metric");
  if (metric_attnum > 0) {
    values[metric_attnum - 1] = CStringGetTextDatum(state->metric.name);
    nulls[metric_attnum - 1] = false;
  }

  return heap_form_tuple(tupdesc, values, nulls);
}

/**
 * Parse an influx packet.
 */
Datum parse_influx(PG_FUNCTION_ARGS) {
  HeapTuple tuple;
  FuncCallContext *funcctx;
  IngestState *state;

  if (SRF_IS_FIRSTCALL()) {
    TupleDesc tupdesc;
    MemoryContext oldcontext;

    funcctx = SRF_FIRSTCALL_INIT();
    oldcontext = MemoryContextSwitchTo(funcctx->multi_call_memory_ctx);

    if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
      ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
                      errmsg("function returning record called in context "
                             "that cannot accept type record")));
    funcctx->attinmeta = TupleDescGetAttInMetadata(tupdesc);
    funcctx->user_fctx =
        ParseInfluxSetup(text_to_cstring(PG_GETARG_TEXT_PP(0)));

    MemoryContextSwitchTo(oldcontext);
  }

  funcctx = SRF_PERCALL_SETUP();
  state = funcctx->user_fctx;

  tuple = ParseInfluxNextTuple(state, funcctx->attinmeta);
  if (tuple == NULL)
    SRF_RETURN_DONE(funcctx);

  SRF_RETURN_NEXT(funcctx, HeapTupleGetDatum(tuple));
}

static void StartBackgroundWorkers(const char *database_name,
                                   const char *schema_name,
                                   const char *role_name,
                                   const char *service_name) {
  MemoryContext oldcontext = MemoryContextSwitchTo(TopMemoryContext);
  WorkerArgs args = {0};

  if (schema_name)
    strncpy(args.namespace, schema_name, sizeof(args.namespace));
  if (database_name)
    strncpy(args.database, database_name, sizeof(args.database));
  if (role_name)
    strncpy(args.role, role_name, sizeof(args.role));
  if (service_name)
    strncpy(args.service, service_name, sizeof(args.service));

  StartUdpBackgroundWorkers(&args);
  MemoryContextSwitchTo(oldcontext);
}

void _PG_init(void) {
  InfluxGucInitDynamic();

  if (!process_shared_preload_libraries_in_progress)
    return;

  InfluxGucInitStatic();
  StartBackgroundWorkers(InfluxDatabaseName, InfluxSchemaName, InfluxRoleName,
                         InfluxServiceName);
}
