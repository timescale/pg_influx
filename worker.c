/*
 * Copyright 2022 Timescale Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "worker.h"

#include <postgres.h>

#include <access/xact.h>
#include <catalog/namespace.h>
#include <pgstat.h>
#include <storage/latch.h>
#include <utils/elog.h>
#if PG_VERSION_NUM < 150000
#include <utils/int8.h>
#endif

#include <stdbool.h>
#include <string.h>
#include <unistd.h>

#include "cache.h"
#include "influx.h"

volatile sig_atomic_t ReloadConfig = false;
volatile sig_atomic_t ShutdownWorker = false;

/**
 * Namespace OID for schema to write metrics to. This is shared by all
 * kinds of protocols.
 */
Oid NamespaceOid;

/* Check that sizeof(WorkerArgs) > BGW_EXTRALEN */
static char c1[BGW_EXTRALEN - sizeof(WorkerArgs)] pg_attribute_unused();

/**
 * Process one packet of lines.
 *
 * The packet consists of zero or more lines of Influx line protocol
 * lines. These will be read and inserted into the database under the
 * apropriate metric.
 */
void InfluxProcessPacket(char *buffer, size_t bytes) {
  IngestState *state;

  buffer[bytes] = '\0';
  state = ParseInfluxSetup(buffer);

  while (true) {
    bool result;
    PG_TRY();
    { result = IngestReadNextLine(state); }
    PG_CATCH();
    { result = false; }
    PG_END_TRY();
    if (!result)
      return;
    MetricInsert(&state->metric);
  }
}

/* Signal handler for SIGTERM */
static void WorkerSigterm(SIGNAL_ARGS) {
  int save_errno = errno;
  ShutdownWorker = true;
  SetLatch(MyLatch);
  errno = save_errno;
}

/* Signal handler for SIGHUP */
static void WorkerSighup(SIGNAL_ARGS) {
  int save_errno = errno;
  ReloadConfig = true;
  SetLatch(MyLatch);
  errno = save_errno;
}

void InfluxWorkerSetup(WorkerArgs *args) {
  /* Establish signal handlers; once that's done, unblock signals. */
  pqsignal(SIGTERM, WorkerSigterm);
  pqsignal(SIGHUP, WorkerSighup);
  BackgroundWorkerUnblockSignals();
  BackgroundWorkerInitializeConnection(args->database, args->role, 0);

  pgstat_report_activity(STATE_RUNNING, "initializing worker");

  CacheInit();

  /* We need to start a transaction first because none is started and
     SPI_connect_ext might use TopTransactionContext, which is set by
     this function. The SPI_commit below will automatically start a
     new one. */
  StartTransactionCommand();

  /* It is necessary to start a transaction before calling functions
     like `get_namespace_oid`. Calling them before this will cause a
     crash. */
  NamespaceOid = get_namespace_oid(args->namespace, false);
}
