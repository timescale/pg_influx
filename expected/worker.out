CREATE SCHEMA db_worker;
CREATE TABLE db_worker.disk(_time timestamptz, host text, device text, _tags jsonb, _fields jsonb);
CREATE TABLE db_worker.system(_time timestamp, host text, uptime int, _tags jsonb, _fields jsonb);
CREATE EXTENSION influx WITH SCHEMA db_worker;
\set VERBOSITY terse
\x on
SELECT pg_sleep(1), pid AS worker_pid FROM db_worker.launch_udp_worker('db_worker', 4711::text) AS pid \gset
CALL db_worker.send_udp_packet('cpu,cpu=cpu0,host=fury usage_system=2.0408163264927324,usage_user=2.0408163264927324 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('cpu,cpu=cpu1,host=fury usage_system=3.921568627286635,usage_user=3.9215686274649673 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('disk,device=nvme0n1p2,fstype=ext4,host=fury,mode=rw,path=/ free=912578965504i,total=1006530654208i,used=42751348736i,used_percent=4.475033200428718 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('disk,device=nvme0n1p1,fstype=vfat,host=fury,mode=rw,path=/boot/efi free=527806464i,total=535805952i,used=7999488i,used_percent=1.4929822952022749 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('system,host=fury load1=2.13,load15=0.84,load5=1.18,n_cpus=8i,n_users=1i 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('system,host=fury uptime=607641i 1574753954000000000', 4711::text);
CALL db_worker.send_udp_packet('system,host=fury uptime_format="7 days,  0:47" 1574753954000000000', 4711::text);
SELECT pg_sleep(2);
-[ RECORD 1 ]
pg_sleep | 

SELECT * FROM db_worker.cpu;	--Automatically created
-[ RECORD 1 ]-----------------------------------------------------------------------
_time   | Mon Nov 25 23:39:14 2019 PST
_tags   | {"cpu": "cpu0", "host": "fury"}
_fields | {"usage_user": "2.0408163264927324", "usage_system": "2.0408163264927324"}
-[ RECORD 2 ]-----------------------------------------------------------------------
_time   | Mon Nov 25 23:39:14 2019 PST
_tags   | {"cpu": "cpu1", "host": "fury"}
_fields | {"usage_user": "3.9215686274649673", "usage_system": "3.921568627286635"}

SELECT * FROM db_worker.disk;
-[ RECORD 1 ]-----------------------------------------------------------------------------------------------------------
_time   | Mon Nov 25 23:39:14 2019 PST
host    | fury
device  | nvme0n1p2
_tags   | {"mode": "rw", "path": "/", "fstype": "ext4"}
_fields | {"free": "912578965504", "used": "42751348736", "total": "1006530654208", "used_percent": "4.475033200428718"}
-[ RECORD 2 ]-----------------------------------------------------------------------------------------------------------
_time   | Mon Nov 25 23:39:14 2019 PST
host    | fury
device  | nvme0n1p1
_tags   | {"mode": "rw", "path": "/boot/efi", "fstype": "vfat"}
_fields | {"free": "527806464", "used": "7999488", "total": "535805952", "used_percent": "1.4929822952022749"}

SELECT * FROM db_worker.system;
-[ RECORD 1 ]--------------------------------------------------------------------------------
_time   | Tue Nov 26 07:39:14 2019
host    | fury
uptime  | 
_tags   | {}
_fields | {"load1": "2.13", "load5": "1.18", "load15": "0.84", "n_cpus": "8", "n_users": "1"}
-[ RECORD 2 ]--------------------------------------------------------------------------------
_time   | Tue Nov 26 07:39:14 2019
host    | fury
uptime  | 607641
_tags   | {}
_fields | {}
-[ RECORD 3 ]--------------------------------------------------------------------------------
_time   | Tue Nov 26 07:39:14 2019
host    | fury
uptime  | 
_tags   | {}
_fields | {"uptime_format": "7 days,  0:47"}

SELECT count(*) FROM pg_stat_activity WHERE pid = :worker_pid;
-[ RECORD 1 ]
count | 1

-- Syntax error, but the worker should not stop
CALL db_worker.send_udp_packet('system,host=fury', 4711::text);
SELECT pg_sleep(1);
-[ RECORD 1 ]
pg_sleep | 

SELECT count(*) FROM pg_stat_activity WHERE pid = :worker_pid;
-[ RECORD 1 ]
count | 1

SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE backend_type like '%Influx%';
-[ RECORD 1 ]--------+--
pg_terminate_backend | t

DROP EXTENSION influx;
DROP TABLE db_worker.cpu;
DROP TABLE db_worker.disk;
DROP TABLE db_worker.system;
DROP SCHEMA db_worker;
